### PR TITLE
Display updated results

### PR DESCRIPTION
**Pull request by hrsh7th**

### CODE REVIEW

File: autoload/jetpack.vim [hunk 1]
```diff
@@ -14,23 +14,28 @@ let s:srcdir = s:home .. '/pack/jetpack/src'
...
+\   '**/README*',
+\   '**/Rakefile*',
+\   '**/Gemfile*',
+\   '**/Makefile*',
+\   '**/LICENSE*',
+\   '**/LICENCE*',
+\   '**/CONTRIBUTING*',
+\   '**/CHANGELOG*',
+\   '**/NEWS*',
+\ ]
```
Comment (hrsh7th) [inline autoload/jetpack.vim:32]: indent only change.

File: autoload/jetpack.vim [hunk 2]
```diff
@@ -92,20 +106,18 @@ function s:setbufline(lnum, text, ...)
   redraw
 endfunction
 
-function s:createbuf()
-  silent vsplit +setlocal\ buftype=nofile\ nobuflisted\ noswapfile\ nonumber\ nowrap JetpackStatus
+function s:setupbuf()
+  if bufexists('JetpackStatus')
+    execute 'bdelete! ' .. bufnr('JetpackStatus')
+  endif
+  silent vnew +setlocal\ buftype=nofile\ nobuflisted\ noswapfile\ nonumber\ nowrap JetpackStatus
```
Comment (hrsh7th) [inline autoload/jetpack.vim:113]: Use vnew

File: autoload/jetpack.vim [hunk 3]
```diff
@@ -116,38 +128,53 @@ function jetpack#install(...)
...
-    let job = s:jobstart(cmd, function('<SID>setbufline', [i+3, printf('Installed %s', pkg.name)]))
+    let job = s:jobstart(cmd, function({ i, pkg, output -> [
+    \   extend(pkg, {
+    \     'progress': {
+    \       'type': s:progress_type.install,
+    \       'output': output
+    \     }
+    \   }),
+    \   s:setbufline(i+3, printf('Installed %s', pkg.name))
+    \ ] }, [i, pkg]))
```
Comment (hrsh7th) [inline autoload/jetpack.vim:145]: Append install results to each pkg.

File: autoload/jetpack.vim [hunk 4]
```diff
@@ -116,38 +128,53 @@ function jetpack#install(...)
...
+  call s:setupbuf()
   let jobs = []
   for i in range(len(s:pkgs))
     let pkg = s:pkgs[i]
     call s:setbufline(1, printf('Update Plugins (%d / %d)', (len(jobs) - s:jobcount(jobs)), len(s:pkgs)))
     call s:setbufline(2, s:progressbar((0.0 + len(jobs) - s:jobcount(jobs)) / len(s:pkgs) * 100))
     call s:setbufline(i+3, printf('Updating %s ...', pkg.name))
-    if (a:0 > 0 && index(a:000, pkg.name) < 0) || (pkg.frozen || !isdirectory(pkg.path))
+    if pkg.progress.type ==# s:progress_type.install || (a:0 > 0 && index(a:000, pkg.name) < 0) || (pkg.frozen || !isdirectory(pkg.path))
       call s:setbufline(i+3, printf('Skipped %s', pkg.name))
```
Comment (hrsh7th) [inline autoload/jetpack.vim:161]: Skip installed plugins.

File: autoload/jetpack.vim [hunk 5]
```diff
@@ -116,38 +128,53 @@ function jetpack#install(...)
...
+    let cmd = ['git', '-C', pkg.path, 'pull', '--rebase']
+    let job = s:jobstart(cmd, function({ i, pkg, output -> [
+    \   extend(pkg, {
+    \     'progress': {
+    \       'type': s:progress_type.update,
+    \       'output': output
+    \     }
+    \   }),
+    \   s:setbufline(i+3, printf('Updated %s', pkg.name))
+    \ ] }, [i, pkg]))
```
Comment (hrsh7th) [inline autoload/jetpack.vim:173]: Append updated results to each pkg.

File: autoload/jetpack.vim [hunk 6]
```diff
@@ -172,44 +201,69 @@ function jetpack#bundle()
...
-    call s:setbufline(1, printf('Merging Plugins (%d / %d)', i, len(s:pkgs)))
-    call s:setbufline(2, s:progressbar(1.0 * i / len(s:pkgs) * 100))
-    call s:setbufline(i+3, printf('Merging %s ...', pkg.name))
+    call s:setbufline(1, printf('Merging Plugins (%d / %d)', merged_count, len(s:pkgs)))
+    call s:setbufline(2, s:progressbar(1.0 * merged_count / len(s:pkgs) * 100))
     let srcdir = pkg.path .. '/' .. pkg.subdir
     let srcfiles = filter(s:files(srcdir), "!s:ignorable(substitute(v:val, srcdir, '', ''))")
     let destfiles = map(copy(srcfiles), "substitute(v:val, srcdir, destdir, '')")
-    let dupfiles = filter(copy(destfiles), "filereadable(v:val)")
+    let dupfiles = filter(copy(destfiles), "has_key(merged_files, v:val)")
```
Comment (hrsh7th) [inline autoload/jetpack.vim:214]: Avoid calling filereadable

File: autoload/jetpack.vim [hunk 7]
```diff
@@ -172,44 +201,69 @@ function jetpack#bundle()
...
+  let msg[s:progress_type.install] = 'Installed'
+  let msg[s:progress_type.update] = 'Updated'
+
+  let line_count = 1
+  for pkg in s:pkgs
+    call s:setbufline(line_count, printf('%s %s', msg[pkg.progress.type], pkg.name))
+    let line_count += 1
+
+    let output = pkg.progress.output
+    let output = substitute(output, '^From.\{-}\zs\n\s*', '/compare/', 'g')
```
Comment (hrsh7th) [inline autoload/jetpack.vim:256]: Hack. Create compare url.

File: autoload/jetpack.vim [hunk 8]
```diff
@@ -162,43 +201,69 @@ function jetpack#bundle()
...
       let destfile = substitute(srcfile, srcdir, destdir, '')
-      call mkdir(fnamemodify(destfile, ':p:h'), 'p')
-      call s:copy(srcfile, destfile)
+      call s:copy(srcfile, substitute(srcfile, srcdir, destdir, ''))
+    endfor
+    call s:setbufline(i+merged_count+3, printf('Copied %s ...', pkg.name))
+  endfor
+endfunction
+
+function jetpack#diplay()
```
Comment (tani) [inline autoload/jetpack.vim:243]: As `jetpack#display` will never be called out of the script, how about `s:display`?

Comment (hrsh7th) [reply autoload/jetpack.vim:243]: Yes. It can be changed because it is only referencing the convention such as jetpack#update etc.

**Comment by tani** [comment]
Thanks for the awesome UX improvement! I am really excited to merge this.

**Pull request merged**